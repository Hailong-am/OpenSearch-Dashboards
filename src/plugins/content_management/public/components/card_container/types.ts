/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiCardSelectProps } from '@elastic/eui/src/components/card/card_select';
import { ContainerInput } from '../../../../embeddable/public';

export interface CardExplicitInput {
  title: string;
  description: string;
  onClick?: () => void;
  getIcon?: () => React.ReactElement;
  getFooter?: () => React.ReactElement;
  selectable?: EuiCardSelectProps;
}

export type CardContainerInput = ContainerInput<CardExplicitInput> & {
  columns?: number;
  wrap?: boolean;
};

/**
 * The props which allow to be updated after card container was created
 */
export type CardContainerExplicitInput = Partial<Pick<CardContainerInput, 'title' | 'columns'>>;
