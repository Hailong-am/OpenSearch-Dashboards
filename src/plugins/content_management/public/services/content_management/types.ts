/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiCardSelectProps } from '@elastic/eui/src/components/card/card_select';
import { CardContainerExplicitInput } from '../../components/card_container/types';
import { DashboardContainerExplicitInput } from '../../components/types';

export interface PageConfig {
  id: string;
  title?: string;
  description?: string;
  sections?: Section[];
}

export type Section =
  | {
      kind: 'custom';
      id: string;
      order: number;
      title?: string;
      description?: string;
      render: (contents: Content[]) => JSX.Element;
    }
  | {
      kind: 'dashboard';
      id: string;
      order: number;
      title?: string;
      description?: string;
      input?: DashboardContainerExplicitInput;
    }
  | {
      kind: 'card';
      id: string;
      order: number;
      title?: string;
      columns?: number;
      input?: CardContainerExplicitInput;
    };

export type Content =
  | {
      kind: 'visualization';
      id: string;
      order: number;
      input: SavedObjectInput;
    }
  | {
      kind: 'dashboard';
      id: string;
      order: number;
      input: SavedObjectInput;
      width: number;
    }
  | {
      kind: 'custom';
      id: string;
      order: number;
      render: () => JSX.Element;
    }
  | {
      kind: 'card';
      id: string;
      order: number;
      title: string;
      description: string;
      onClick?: () => void;
      getIcon?: () => React.ReactElement;
      getFooter?: () => React.ReactElement;
      selectable?: EuiCardSelectProps;
    };

export type SavedObjectInput =
  | {
      kind: 'static';
      /**
       * The visualization id
       */
      id: string;
    }
  | {
      kind: 'dynamic';
      /**
       * A promise that returns a visualization id
       */
      get: () => Promise<string>;
    };

export interface ContentProvider {
  id: string;
  getContent: () => Content;
  getTargetArea: () => string | string[];
}
