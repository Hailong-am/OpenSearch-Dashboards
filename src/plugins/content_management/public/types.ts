/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { CoreStart } from 'opensearch-dashboards/public';

import { ContentManagementService, ContentProvider } from './services';
import { EmbeddableSetup, EmbeddableStart } from '../../embeddable/public';

export interface ContentManagementPluginSetup {
  registerPage: ContentManagementService['registerPage'];
}
export interface ContentManagementPluginStart {
  registerContentProvider: (provider: ContentProvider) => void;
  renderPage: (id: string) => React.ReactNode;
}

export type ContentManagementPluginStartDependencies = {
  embeddable: EmbeddableStart;
};

export type ContentManagementPluginSetupDependencies = {
  embeddable: EmbeddableSetup;
};

export type ContentServices = CoreStart & ContentManagementPluginStartDependencies;
