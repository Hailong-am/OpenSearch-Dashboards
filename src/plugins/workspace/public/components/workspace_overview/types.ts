/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { AppCategory } from '../../../../../core/types';

export interface GetStartCard {
  /**
   * application id that used for filter, if app id is not specified, that means it will always display
   */
  appId?: string;
  /**
   * feature Name
   */
  featureName: string;
  /**
   * card description
   */
  featureDescription: string;
  /**
   * redirect destination when the card been clicked
   */
  link?: string;
  /**
   * application category which appId belongs to
   */
  category: AppCategory;
  /**
   * order of card
   */
  order?: 0;
}
