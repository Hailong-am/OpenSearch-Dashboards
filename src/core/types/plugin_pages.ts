/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export interface PluginPages {
  title: string;
  url: string;
  order: number;
}
