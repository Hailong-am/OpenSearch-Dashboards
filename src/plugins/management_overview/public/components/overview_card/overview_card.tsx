/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiHorizontalRule,
  EuiListGroup,
  EuiListGroupItem,
  EuiPanel,
  EuiTitle,
} from '@elastic/eui';
import React from 'react';
import { PluginPages } from 'src/core/types/plugin_pages';

export interface OverviewCardProps {
  title: string;
  pages: PluginPages[];
  onClick: (url: string) => void;
}

const MAX_ITEM = 4;

export function OverviewCard(props: OverviewCardProps) {
  const { title, pages, onClick } = props;

  const showViewMore = pages && pages.length > MAX_ITEM;
  let pagesToShow = pages;
  if (showViewMore) {
    pagesToShow = pages.slice(0, MAX_ITEM - 1);
    pagesToShow.push({
      title: 'View more...',
      url: '',
      order: 1000,
    });
  }

  return (
    <EuiPanel>
      <EuiTitle size="s">
        <h5>{title}</h5>
      </EuiTitle>
      <EuiHorizontalRule margin="s" />
      <EuiListGroup gutterSize="none" size="s">
        {pagesToShow.map((page) => (
          <EuiListGroupItem
            key={page.title}
            onClick={() => onClick(page.url)}
            label={page.title}
            color="primary"
            size="s"
          />
        ))}
      </EuiListGroup>
    </EuiPanel>
  );
}
