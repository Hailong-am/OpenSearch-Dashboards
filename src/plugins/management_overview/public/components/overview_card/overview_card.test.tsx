/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { render } from '@testing-library/react';
import { OverviewCard } from './overview_card';
import React from 'react';
import { PluginPages } from 'src/core/types';

function renderOverviewCard(pages: PluginPages[]) {
  return render(<OverviewCard title="Dev Tools" onClick={jest.fn()} pages={pages} />);
}

describe('OverviewCard', () => {
  const singlePage: PluginPages[] = [
    {
      title: 'Dev tools',
      order: 100,
      url: '#/console',
    },
  ];

  const multiplePages: PluginPages[] = [
    {
      title: 'Dev tools-1',
      order: 100,
      url: '#/page1',
    },
    {
      title: 'Dev tools-2',
      order: 200,
      url: '#/page2',
    },
    {
      title: 'Dev tools-3',
      order: 300,
      url: '#/page3',
    },
    {
      title: 'Dev tools-4',
      order: 400,
      url: '#/page1',
    },
    {
      title: 'Dev tools-5',
      order: 500,
      url: '#/page5',
    },
  ];

  it('should render normally', () => {
    const { container, queryByText } = renderOverviewCard(singlePage);
    expect(container.firstChild).toMatchSnapshot();
    expect(queryByText('View more...')).toBeNull();
  });

  it('should render normally for more than 4 pages', () => {
    const { container, queryByText } = renderOverviewCard(multiplePages);
    expect(container.firstChild).toMatchSnapshot();
    expect(queryByText('View more...')).not.toBeNull();
  });
});
