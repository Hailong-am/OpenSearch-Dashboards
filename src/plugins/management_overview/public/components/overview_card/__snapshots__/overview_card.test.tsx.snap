// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewCard should render normally 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
>
  <h5
    class="euiTitle euiTitle--small"
  >
    Dev Tools
  </h5>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
  />
  <ul
    class="euiListGroup euiListGroup-maxWidthDefault"
  >
    <li
      class="euiListGroupItem euiListGroupItem--small euiListGroupItem--primary euiListGroupItem-isClickable"
    >
      <button
        class="euiListGroupItem__button"
        type="button"
      >
        <span
          class="euiListGroupItem__label"
          title="Dev tools"
        >
          Dev tools
        </span>
      </button>
    </li>
  </ul>
</div>
`;

exports[`OverviewCard should render normally for more than 4 pages 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
>
  <h5
    class="euiTitle euiTitle--small"
  >
    Dev Tools
  </h5>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
  />
  <ul
    class="euiListGroup euiListGroup-maxWidthDefault"
  >
    <li
      class="euiListGroupItem euiListGroupItem--small euiListGroupItem--primary euiListGroupItem-isClickable"
    >
      <button
        class="euiListGroupItem__button"
        type="button"
      >
        <span
          class="euiListGroupItem__label"
          title="Dev tools-1"
        >
          Dev tools-1
        </span>
      </button>
    </li>
    <li
      class="euiListGroupItem euiListGroupItem--small euiListGroupItem--primary euiListGroupItem-isClickable"
    >
      <button
        class="euiListGroupItem__button"
        type="button"
      >
        <span
          class="euiListGroupItem__label"
          title="Dev tools-2"
        >
          Dev tools-2
        </span>
      </button>
    </li>
    <li
      class="euiListGroupItem euiListGroupItem--small euiListGroupItem--primary euiListGroupItem-isClickable"
    >
      <button
        class="euiListGroupItem__button"
        type="button"
      >
        <span
          class="euiListGroupItem__label"
          title="Dev tools-3"
        >
          Dev tools-3
        </span>
      </button>
    </li>
    <li
      class="euiListGroupItem euiListGroupItem--small euiListGroupItem--primary euiListGroupItem-isClickable"
    >
      <button
        class="euiListGroupItem__button"
        type="button"
      >
        <span
          class="euiListGroupItem__label"
          title="View more..."
        >
          View more...
        </span>
      </button>
    </li>
  </ul>
</div>
`;
